File: js/shape-parts.js
/** @jsx React.DOM */

/*
 * Copyright 2013 Michael M. Magruder (https://github.com/mikemag)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var NoseShapeSimpleBezier = React.createClass({

    getInitialState: function() {
        return {pullStart: this.props.board.noseWidth / 3,
                pullEnd: this.props.board.noseLength / 3};
    },

    render: function() {
        this.state.pullStart = this.props.board.noseWidth / 3;
        this.state.pullEnd = this.props.board.noseLength / 3;
        this.draw();
        return (
                <div>
                  <Dimension name="Nose length">
                    {this.props.board.noseLength}
                  </Dimension>
                  <Dimension name="Nose width">
                    {this.props.board.noseWidth}
                  </Dimension>
                  <Dimension name="Pull start">{this.state.pullStart}</Dimension>
                  <Dimension name="Pull end">{this.state.pullEnd}</Dimension>
                </div>
        );
    },

    draw: function() {
        var b = this.props.board;
        var context = startDesignDrawing(context);
        context.beginPath();
        context.moveTo(0, 0);
        context.bezierCurveTo(0, this.state.pullStart,
                              this.state.pullEnd, b.noseWidth / 2,
                              b.noseLength, b.noseWidth / 2);
        context.moveTo(0, 0);
        context.bezierCurveTo(0, -this.state.pullStart,
                              this.state.pullEnd, b.noseWidth / -2,
                              b.noseLength, b.noseWidth / -2);
        context.lineWidth = 0.5;
        context.strokeStyle = 'white';
        context.stroke();
        endDesignDrawing(context);
    },

    getName: function() {
        return "Simple Bezier";
    },

    getDefinition: function() {
        // @todo: return the definition of this part to be sent to the server
        // for accurate rendering.
    },
});

var NoseShapeFatBezier = React.createClass({

    getInitialState: function() {
        return {pullStart: this.props.board.noseWidth / 1.8,
                pullEnd: this.props.board.noseLength / 1.8};
    },

    render: function() {
        this.state.pullStart = this.props.board.noseWidth / 1.8;
        this.state.pullEnd = this.props.board.noseLength / 1.8;
        this.draw();
        return (
                <div>
                  <Dimension name="Nose length">
                    {this.props.board.noseLength}
                  </Dimension>
                  <Dimension name="Nose width">
                    {this.props.board.noseWidth}
                  </Dimension>
                  <Dimension name="Pull start">{this.state.pullStart}</Dimension>
                  <Dimension name="Pull end">{this.state.pullEnd}</Dimension>
                </div>
        );
    },

    draw: function() {
        var b = this.props.board;
        var context = startDesignDrawing(context);
        context.beginPath();
        context.moveTo(0, 0);
        context.bezierCurveTo(0, this.state.pullStart,
                              this.state.pullEnd, b.noseWidth / 2,
                              b.noseLength, b.noseWidth / 2);
        context.moveTo(0, 0);
        context.bezierCurveTo(0, -this.state.pullStart,
                              this.state.pullEnd, b.noseWidth / -2,
                              b.noseLength, b.noseWidth / -2);
        context.lineWidth = 0.5;
        context.strokeStyle = 'white';
        context.stroke();
        endDesignDrawing(context);
    },

    getName: function() {
        return "Fat Bezier";
    },

    getDefinition: function() {
        // @todo: return the definition of this part to be sent to the server
        // for accurate rendering.
    },
});

var TailShapeSimpleBezier = React.createClass({

    getInitialState: function() {
        return {pullStart: this.props.board.tailWidth / 3,
                pullEnd: this.props.board.tailLength / 3};
    },

    render: function() {
        this.state.pullStart = this.props.board.tailWidth / 3;
        this.state.pullEnd = this.props.board.tailLength / 3;
        this.draw();
        return (
                <div>
                  <Dimension name="Tail length">
                    {this.props.board.tailLength}
                  </Dimension>
                  <Dimension name="Tail width">
                    {this.props.board.tailWidth}
                  </Dimension>
                  <Dimension name="Pull start">{this.state.pullStart}</Dimension>
                  <Dimension name="Pull end">{this.state.pullEnd}</Dimension>
                </div>
        );
    },

    draw: function() {
        var b = this.props.board;
        var context = startDesignDrawing(context);
        context.beginPath();
        context.moveTo(b.overallLength, 0);
        var tailX = b.noseLength + b.eeLength;
        context.bezierCurveTo(b.overallLength, this.state.pullStart,
                              b.overallLength - this.state.pullEnd,
                              b.tailWidth / 2,
                              tailX, b.tailWidth / 2);
        context.moveTo(b.overallLength, 0);
        context.bezierCurveTo(b.overallLength, -this.state.pullStart,
                              b.overallLength - this.state.pullEnd,
                              b.tailWidth / -2,
                              tailX, b.tailWidth / -2);
        context.lineWidth = 0.5;
        context.strokeStyle = 'white';
        context.stroke();
        endDesignDrawing(context);
    },

    getName: function() {
        return "Simple Bezier";
    },

    getDefinition: function() {
        // @todo: return the definition of this part to be sent to the server
        // for accurate rendering.
    },
});

var EdgeShapeSimpleArc = React.createClass({

    getInitialState: function() {
        return {sidecutRadius: this.props.board.sidecutRadius,
                eeLength: this.props.board.eeLength};
    },

    render: function() {
        this.state.sidecutRadius = this.props.board.sidecutRadius;
        this.state.eeLength = this.props.board.eeLength;
        this.draw();
        return (
                <div>
                  <Dimension name="EE length">
                    {this.props.board.eeLength}
                  </Dimension>
                  <Dimension name="Waist width">
                    {this.props.board.waistWidth}
                  </Dimension>
                  <Dimension name="Taper" unit="mm">
                    {this.props.board.taper}
                  </Dimension>
                  <Dimension name="Sidecut radius">
                    {this.state.sidecutRadius}
                  </Dimension>
                  <Dimension name="Sidecut depth">
                    {this.props.board.sidecutDepth}
                  </Dimension>
                </div>
        );
    },

    draw: function() {
        var b = this.props.board;
        edgeDesignShape(
            b.noseLength,
            b.noseWidth / -2,
            b.noseLength + (b.eeLength / 2),
            b.waistWidth / -2,
            b.noseLength + b.eeLength,
            b.tailWidth / -2,
            true
        );
        edgeDesignShape(
            b.noseLength,
            b.noseWidth / 2,
            b.noseLength + (b.eeLength / 2),
            b.waistWidth / 2,
            b.noseLength + b.eeLength,
            b.tailWidth / 2,
            false
        );
    },

    getName: function() {
        return "Simple Arc";
    },

    getDefinition: function() {
        // @todo: return the definition of this part to be sent to the server
        // for accurate rendering.
    },
});

File: core/shape-parts.cpp
/*
 * Copyright 2013-2014 Michael M. Magruder (https://github.com/mikemag)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "shape-parts.h"

namespace MonkeyCAM {

const Path& BasicBezier::generate(Point endPoint, Point transitionPoint) {
  if (endPoint.X < transitionPoint.X) {
    m_path = BezierPath(endPoint,
                        Point(endPoint.X, (transitionPoint.Y - endPoint.Y) *
                              m_endHandle),
                        Point(transitionPoint.X -
                              ((transitionPoint.X - endPoint.X) *
                               m_transitionHandle), transitionPoint.Y),
                        transitionPoint);
  } else {
    m_path = BezierPath(transitionPoint,
                        Point(transitionPoint.X +
                              ((endPoint.X - transitionPoint.X) *
                               m_transitionHandle), transitionPoint.Y),
                        Point(endPoint.X, (transitionPoint.Y - endPoint.Y) *
                              m_endHandle),
                        endPoint);
  }
  return m_path;
}

const Path& FlatBezier::generate(Point endPoint, Point transitionPoint) {
  MCFixed flatEndY = - m_flatWidth / 2; 
  // Note: flatEndY is negative because the 'bottom' half of the board is drawn 
  //       first (i.e. in negative Y-space)
  BOOST_ASSERT_MSG(flatEndY >= transitionPoint.Y, 
                   "Error in generating \"Flat\" end: \"flat width\" is wider than nose/tail width");
  if (endPoint.X < transitionPoint.X) { 
    // Nose
    Path ep;
    ep.push_back(endPoint);
    m_path = BezierPath(Point(endPoint.X, flatEndY),
                        Point(endPoint.X, flatEndY + 
                              (transitionPoint.Y - flatEndY) * m_endHandle),
                        Point(transitionPoint.X -
                              ((transitionPoint.X - endPoint.X) *
                               m_transitionHandle), transitionPoint.Y),
                        transitionPoint);
    ep.push_back_path(m_path);
    m_path = ep;
  } else {
    // Tail
    m_path = BezierPath(transitionPoint,
                        Point(transitionPoint.X +
                              ((endPoint.X - transitionPoint.X) *
                               m_transitionHandle), transitionPoint.Y),
                        Point(endPoint.X, flatEndY + 
                              (transitionPoint.Y - flatEndY) * m_endHandle),
                        Point(endPoint.X, flatEndY));
    m_path.push_back(endPoint);
  }
  return m_path;
}
const Path& BasicArc::generate(Point nosePoint, Point waistPoint,
                               Point tailPoint) {
  m_path = ArcPath { nosePoint, waistPoint, tailPoint, ArcPath::Clockwise };
  return m_path;
}

} // namespace MonkeyCAM
